presto-main/src/main/java/com/facebook/presto/execution/SqlTaskExecution.java

```diff
@@ -1118,18 +1141,22 @@ public synchronized void decrementPendingCreation(int pipelineId, Lifespan lifes
             per(pipelineId).pendingCreation--;
         }
 
-        public synchronized void incrementRemainingDriver(Lifespan lifespan)
+        public synchronized void incrementRemainingDriver(Lifespan lifespan, boolean outputDriver)
         {
             checkState(!isNoMoreDriverRunners(lifespan), "Cannot increment remainingDriver for Lifespan %s. NoMoreSplits is set.", lifespan);
             per(lifespan).remainingDriver++;
-            overallRemainingDriver++;
+            if (outputDriver) {
+                remainingOutputDriver++;
+            }
         }
 
-        public synchronized void decrementRemainingDriver(Lifespan lifespan)
+        public synchronized void decrementRemainingDriver(Lifespan lifespan, boolean outputDriver)
         {
             checkState(per(lifespan).remainingDriver > 0, "Cannot decrement remainingDriver for Lifespan %s. Value is 0.", lifespan);
             per(lifespan).remainingDriver--;
-            overallRemainingDriver--;
+            if (outputDriver) {
+                remainingOutputDriver--;
+            }
             checkLifespanCompletion(lifespan);
```

Comment:
A similar set of changes should apply to lifespan:

* It should tracking remainingOutputDriver instead
* A lifespan should early terminate when all its output drivers are completed (and there will be no more), just like at the task level.

Actually, if you do it at the lifespan level, you wouldn't need to worry about individual ones.